dm/pkg/exstorage/utils.go

```diff
@@ -0,0 +1,152 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package exstorage
+
+import (
+	"context"
+	"net/url"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/br/pkg/storage"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+// AdjustPath adjust rawURL, add uniqueId as path suffix.
+func AdjustPath(rawURL string, uniqueID string) (string, error) {
+	if rawURL == "" {
+		return rawURL, nil
+	}
+	u, err := storage.ParseRawURL(rawURL)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	trimPath := strings.TrimRight(u.Path, "/")
+	// avoid duplicate add uniqueID
+	if uniqueID != "" && !strings.HasSuffix(trimPath, uniqueID) {
+		u.Path = trimPath + "." + uniqueID
+		newURL, err := url.QueryUnescape(u.String())
+		if err != nil {
+			return "", errors.Trace(err)
+		}
+		return newURL, nil
+	}
+
+	return rawURL, nil
+}
+
+// isS3Path judges if rawURL is s3 path.
+func IsS3Path(rawURL string) (bool, error) {
+	if rawURL == "" {
+		return false, nil
+	}
+	u, err := url.Parse(rawURL)
+	if err != nil {
+		return false, err
+	}
+	if u.Scheme == "s3" {
+		return true, nil
+	}
+	return false, nil
+}
+
+// CreateExternalStore creates ExternalStore.
+func CreateExternalStore(ctx context.Context, path string) (storage.ExternalStorage, error) {
+	backend, err := storage.ParseBackend(path, nil)
+	if err != nil {
+		return nil, err
+	}
+	return storage.New(ctx, backend, &storage.ExternalStorageOptions{})
+}
+
+// CollectDirFiles gets files in dir.
+func CollectDirFiles(ctx context.Context, dir string, externalStore storage.ExternalStorage) (map[string]struct{}, error) {
+	var err error
+	if externalStore == nil {
+		externalStore, err = CreateExternalStore(ctx, dir)
+		if err != nil {
+			return nil, err
+		}
+	}
+	files := make(map[string]struct{})
+
+	err = externalStore.WalkDir(ctx, &storage.WalkOption{ListCount: 1}, func(filePath string, size int64) error {
+		name := path.Base(filePath)
+		files[name] = struct{}{}
+		return nil
+	})
+
+	return files, err
+}
+
+// RemoveAll remove files in dir.
+func RemoveAll(ctx context.Context, dir string, externalStore storage.ExternalStorage) error {
+	var err error
+	if externalStore == nil {
+		externalStore, err = CreateExternalStore(ctx, dir)
+		if err != nil {
+			return err
+		}
+	}
+
+	err = externalStore.WalkDir(ctx, &storage.WalkOption{ListCount: 1}, func(filePath string, size int64) error {
+		return externalStore.DeleteFile(ctx, filePath)
+	})
+	if err == nil {
+		return externalStore.DeleteFile(ctx, "")
+	}
+	return err
+}
+
+func ReadFile(ctx context.Context, dir, fileName string, externalStore storage.ExternalStorage) ([]byte, error) {
+	var err error
+	if externalStore == nil {
+		externalStore, err = CreateExternalStore(ctx, dir)
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
a lot of api has this part of code, should we wrap it into a type, so we can just `store.Read` and `store.Open`, etc..

Reply:
`externalStore` may be the only member in new type, and the main purpose of this part of the code is for unit test which need mock a `externalStore`. so we use function may be simpler.